File: LC/Easy/163.py
class Solution:
    def findMissingRanges(self, nums: List[int], lower: int, upper: int) -> List[List[int]]:
        ans = []
        n = len(nums)
        
        if n == 0:
            ans.append([lower, upper])
            return ans
        
        if nums[0] > lower:
            ans.append([lower, nums[0]-1])
        
        for i in range(n - 1):
            if nums[i] + 1 == nums[i+1]:
                continue
                
            ans.append([nums[i]+1, nums[i+1]-1])
            
        if nums[n-1] < upper:
            ans.append([nums[n-1]+1, upper])
            
        return ans

File: LC/C++/Easy/163.cpp
class Solution {
public:
    vector<vector<int>> findMissingRanges(vector<int>& nums, int lower, int upper) {
        std::vector<std::vector<int>> ans;

        if (nums.empty()) {
            return {{lower, upper}};
        }

        if (nums[0] != lower) {
            ans.push_back({lower, nums[0] - 1});
        }

        for (int i = 0; i < nums.size()-1; ++i) {
            if (nums[i] + 1 != nums[i+1]) {
                ans.push_back({nums[i] + 1, nums[i+1] - 1});
            }
        }
        
        if (nums.back() != upper) {
            ans.push_back({nums.back() + 1, upper});
        }

        return ans;
    }
};